pkg/controllers/jobset_controller.go

```diff
@@ -406,6 +407,31 @@ func constructJob(js *jobset.JobSet, rjob *jobset.ReplicatedJob, jobIdx int) (*b
 		Spec: *rjob.Template.Spec.DeepCopy(),
 	}
 
+	// Add labels and annotations to job and pod spec.
+	jobLabelsAndAnnotations(js, rjob, job, jobIdx)
+	podLabelsAndAnnotations(js, rjob, job, jobIdx)
+
+	// If enableDNSHostnames is set, update job spec to set subdomain as
+	// job name (a headless service with same name as job will be created later).
+	if dnsHostnamesEnabled(rjob) {
+		job.Spec.Template.Spec.Subdomain = genSubdomain(js, rjob)
+	}
+
+	// If this job should be exclusive per topology, set the pod affinities/anti-affinities accordingly.
+	if rjob.Exclusive != nil {
+		setExclusiveAffinities(job, rjob.Exclusive.TopologyKey, rjob.Exclusive.NamespaceSelector)
+	}
+
+	return job, nil
+}
+
+func jobLabelsAndAnnotations(js *jobset.JobSet, rjob *jobset.ReplicatedJob, job *batchv1.Job, jobIdx int) {
```

Comment:
There should be a way to unify those two functions into one, perhaps we can pass `[runtime.Object](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/meta.go)`? and use Get/SetLabels and Get/SetAnnotations

also, I would call it `labelAndAnnotateObject`

Reply:
Good idea! Done